src/main/java/org/elasticsearch/plugins/repository/LocalPluginRepository.java

```diff
@@ -0,0 +1,433 @@
+                for (Path pluginDir : root) {
+                    if (Files.isDirectory(pluginDir)) {
+
+                        // Determine the name of the plugin from the directory's name
+                        pluginName = pluginDir.getFileName().toString();
+                        descriptor = pluginDescriptor(pluginName);
+
+                        // Introspect JAR files to find plugin's description
+                        try (DirectoryStream<Path> jarFiles = Files.newDirectoryStream(pluginDir, "*" + JAR)) {
+                            for (Path jarFile : jarFiles) {
+                                // Indicates that the plugin has code/library
+                                descriptor.jvm();
+
+                                String fileName = jarFile.getFileName().toString();
+                                if (!fileName.contains(pluginName)) {
+                                    continue;
+                                }
+
+                                // Constructs the plugin's descriptor from the JAR filename...
+                                // Since plugin's JAR file can have various filenames,
+                                // the result is incertain
+
+                                // Removes "elasticsearch-" prefix
+                                String s = StringUtils.removeStart(fileName, "elasticsearch-");
+
+                                // Removes ".jar" suffix
+                                s = StringUtils.removeEnd(s, JAR);
+
+                                // Removes plugin's name (detected in preVisit() method)
+                                s = StringUtils.removeStart(s, pluginName);
+
+                                // And finally try to extract the version
+                                String version = StringUtils.substringAfterLast(s, "-");
+
+                                if (Strings.hasText(version)) {
+                                    descriptor.version(version);
+                                }
+
+                                // Now, overide everthing with es-plugin.properties or meta.yml file
+                                // contained in the JAR file
+                                descriptor.loadFromZip(this.settings, jarFile);
+                            }
+                        }
+
+                        Path site = pluginDir.resolve(SITE);
+                        if (Files.isDirectory(site)) {
+                            descriptor.site();
+
+                            Path esPlugin = site.resolve(esPluginProps);
+                            if (Files.exists(esPlugin)) {
+                                descriptor.loadFromProperties(esPlugin);
+                            }
+                        }
+
+                        if (descriptor != null) {
+                            PluginDescriptor plugin = descriptor.build();
+                            if ((filter == null) || (filter.accept(plugin))) {
```

Comment:
We never write without any filter. We always write at least a `AlwaysSayYesFilter`... 
So we can simplify this with `if (filter.accept(plugin)) {`